content/blog/fast-data-processing-pipeline-for-predicting-flight-delays-using-apache-apis-kafka-spark-streaming-and-machine-learning-part-2.md

```diff
@@ -0,0 +1,231 @@
+
+![Data Processing Pipeline](/img/data-processing-pipeline.png)
+
+## Spark Streaming Use Case Example Code
+
+The following figure depicts the architecture for the part of the use case data pipeline discussed in this post:
+
+![architecture data pipeline](/img/architechture-data-pipeline.png)
+
+1. ```
+      Flight data is published to a MapR Event Store topic using the Kafka API.
+   ```
+2. ```
+      A Spark streaming application, subscribed to the first topic:
+   ```
+3. ```
+      Ingests a stream of flight data
+   ```
+4. ```
+      Uses a deployed machine learning model to enrich the flight data with a delayed/not delayed prediction
+   ```
+5. ```
+      publishes the results in JSON format to another topic.
+   ```
+6. ```
+      (In the 3rd blog) A Spark streaming application subscribed to the second topic:
+   ```
+7. ```
+      Stores the input data and predictions in MapR Database
+   ```
+
+## Example Use Case Data
+
+You can read more about the data set in part 1 of this series. The incoming and outgoing data is in JSON format, an example is shown below:
+
+```json
+{"_id":"AA_2017-02-16_EWR_ORD_1124”,
+"dofW”:4, "carrier":"AA", "origin":"EWR",
+"dest":"ORD”, "crsdeptime":705,
+"crsarrtime”:851, "crselapsedtime":166.0,"dist":719.0}
+```
+
+![Example Use Case Data](/img/example-use-case-data.png)
+
+## Spark Kafka Consumer Producer Code
+
+**Parsing the Data Set Records**
+
+We use a Scala case class and [Structype](https://spark.apache.org/docs/latest/sql-programming-guide.html#programmatically-specifying-the-schema) to define the schema, corresponding to the input data.
+
+![Structype](/img/structype.png)
+
+**Loading the Model**
+
+The Spark [CrossValidatorModel](https://spark.apache.org/docs/latest/api/scala/index.html#org.apache.spark.ml.tuning.CrossValidatorModel) class is used to load the saved model fitted on the historical flight data.
+
+![Loading the Model](/img/loading-model.png)
+
+**Spark Streaming Code**
+
+These are the basic steps for the Spark Streaming Consumer Producer code:
+
+1. Configure Kafka Consumer and Producer properties.
+2. Initialize a Spark StreamingContext object. Using this context, create a DStream that reads a message from a Topic.
+3. Apply transformations (which create new DStreams).
+4. Write messages from the transformed DStream to a Topic.
+5. Start receiving data and processing. Wait for the processing to be stopped.
+
+**We will go through each of these steps with the example application code.**
+
+## **1) Configure Kafka Consumer Producer properties**
+
+The first step is to set the KafkaConsumer and KafkaProducer configuration properties, which will be used later to create a DStream for receiving/sending messages to topics. You need to set the following paramters:
```

Comment:
Should be spelled “parameters”

![image](https://user-images.githubusercontent.com/77496261/137188178-f912304c-1b6b-4f9b-b698-8ff015342cda.png)

Reply:
done